src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,464 @@
+            // of the out parameters is the same as "this".
+            left = outerBuffer.AsMemory().Slice(ichMin, ichCur - ichMin);
+            right = outerBuffer.AsMemory().Slice(ichCur + 1, ichLim - ichCur - 1);
+            return true;
+        }
+
+        /// <summary>
+        /// Returns a text span with leading and trailing spaces trimmed. Note that this
+        /// will remove only spaces, not any form of whitespace.
+        /// </summary>
+        public static ReadOnlyMemory<char> Trim(ReadOnlyMemory<char> memory)
+        {
+            if (memory.IsEmpty)
+                return memory;
+
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out int ichMin, out int length);
+            int ichLim = ichMin + length;
+            if (outerBuffer[ichMin] != ' ' && outerBuffer[ichLim - 1] != ' ')
+                return memory;
+
+            while (ichMin < ichLim && outerBuffer[ichMin] == ' ')
+                ichMin++;
+            while (ichMin < ichLim && outerBuffer[ichLim - 1] == ' ')
+                ichLim--;
+            return outerBuffer.AsMemory().Slice(ichMin, ichLim - ichMin);
+        }
+
+        /// <summary>
+        /// Returns a text span with leading and trailing whitespace trimmed.
+        /// </summary>
+        public static ReadOnlyMemory<char> TrimWhiteSpace(ReadOnlyMemory<char> memory)
+        {
+            if (memory.IsEmpty)
+                return memory;
+
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out int ichMin, out int length);
+            int ichLim = ichMin + length;
+
+            if (!char.IsWhiteSpace(outerBuffer[ichMin]) && !char.IsWhiteSpace(outerBuffer[ichLim - 1]))
+                return memory;
+
+            while (ichMin < ichLim && char.IsWhiteSpace(outerBuffer[ichMin]))
+                ichMin++;
+            while (ichMin < ichLim && char.IsWhiteSpace(outerBuffer[ichLim - 1]))
+                ichLim--;
+
+            return outerBuffer.AsMemory().Slice(ichMin, ichLim - ichMin);
+        }
+
+        /// <summary>
+        /// Returns a text span with trailing whitespace trimmed.
+        /// </summary>
+        public static ReadOnlyMemory<char> TrimEndWhiteSpace(ReadOnlyMemory<char> memory)
+        {
+            if (memory.IsEmpty)
+                return memory;
+
+            MemoryMarshal.TryGetString(memory, out string outerBuffer, out int ichMin, out int length);
+            int ichLim = ichMin + length;
+            if (!char.IsWhiteSpace(outerBuffer[ichLim - 1]))
+                return memory;
+
+            while (ichMin < ichLim && char.IsWhiteSpace(outerBuffer[ichLim - 1]))
+                ichLim--;
+
+            return outerBuffer.AsMemory().Slice(ichMin, ichLim - ichMin);
+        }
+
+        /// <summary>
+        /// This produces zero for an empty string.
+        /// </summary>
+        public static bool TryParse(out Single value, ReadOnlyMemory<char> memory)
```

Comment:
So, now that .NET has it, is `float.TryParse(ReadOnlyMemory<char>, ...)` unusable?

Reply:
If it is, you can also get rid of `DoubleParser`.

---
In reply to: [212748642](https://github.com/dotnet/machinelearning/pull/705#discussion_r212748642) [](ancestors = 212748642)

Reply:
Very well. Unfortunately, these methods are available in .netstandard 2.1 and our product code targets 2.0. We cannot target 2.1 because then product code won't run on desktop framework anymore because these methods are not available there.

reference: [https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.append?view=netcore-2.1](https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.append?view=netcore-2.1)

[https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.append?view=netframework-4.7.2](https://docs.microsoft.com/en-us/dotnet/api/system.text.stringbuilder.append?view=netframework-4.7.2)


It would have been really nice if they were though...

---
In reply to: [212759741](https://github.com/dotnet/machinelearning/pull/705#discussion_r212759741) [](ancestors = 212759741,212748642)